pyiron_base/generic/filedata.py

```diff
@@ -83,55 +84,78 @@ def load_file(filename, filetype=None, project=None):
 
     """
-    def _load_txt(file):
-        with open(file, encoding='utf8') as f:
-            return f.readlines()
+    def _load_txt(file, is_file):
+        if is_file:
+            with open(file, encoding='utf8') as f:
+                return f.readlines()
+        else:
+            return file.readlines()
 
     def _load_ipynb(file):
         return OwnNotebookNode(nbformat.read(file, as_version=4))
 
-    def _load_json(file):
-        with open(file) as f:
-            return json.load(f)
+    def _load_json(file, is_file):
+        if is_file:
+            with open(file) as f:
+                return json.load(f)
+        else:
+            return json.load(file)
 
     def _load_csv(file):
         return pandas.read_csv(file)
 
     def _load_img(file):
         return Image.open(file)
  
-    def _load_default(file):
+    def _load_default(file, is_file):
         try:
-            return _load_txt(file)
+            return _load_txt(file, is_file)
         except Exception as e:
             raise IOError("File could not be loaded.") from e
 
-    if filetype is None:
-        _, filetype = os.path.splitext(filename)
-    elif filetype[0] != '.':
-        filetype = '.' + filetype
-    filetype = filetype.lower()
+    def _resolve_filetype(file, _filetype):
+        _filename_is_str = isinstance(file, str)
+
+        if _filetype is None and _filename_is_str:
+            _, _filetype = os.path.splitext(file)
+        elif _filetype is None and hasattr(file, 'name'):
+            _, _filetype = os.path.splitext(file.name)
+        elif _filetype is None:
+            pass
+        elif _filetype[0] != '.':
+            _filetype = '.' + _filetype
+        return _filetype.lower(), _filename_is_str
 
-    if filetype in ['.h5', '.hdf']:
+    filetype, filename_is_str = _resolve_filetype(fp, filetype)
+
+    if filetype in ['.h5', '.hdf'] and filename_is_str:
         if project is None:
-            return FileHDFio(file_name=filename)
+            return FileHDFio(file_name=fp)
         else:
-            return ProjectHDFio(file_name=filename, project=project)
+            return ProjectHDFio(file_name=fp, project=project)
     elif filetype in ['.json']:
-        return _load_json(filename)
+        return _load_json(fp, is_file=filename_is_str)
     elif filetype in ['.txt']:
-        return _load_txt(filename)
+        return _load_txt(fp, is_file=filename_is_str)
     elif filetype in ['.csv']:
-        return _load_csv(filename)
+        return _load_csv(fp)
     elif _has_imported['nbformat'] and filetype in ['.ipynb']:
-        return _load_ipynb(filename)
+        return _load_ipynb(fp)
     elif _has_imported['PIL'] and filetype in Image.registered_extensions():
-        return _load_img(filename)
+        return _load_img(fp)
     else:
-        return _load_default(filename)
+        return _load_default(fp, is_file=filename_is_str)
+
+
+class FileDataTemplate(ABC):
```

Comment:
I didn't see it before, but where else is the `FileDataTemplate` supposed to be used?

Reply:
I hope to have it as a Base class for storing File related data. I am using this also for the `S3FileData` [here](https://github.com/pyiron/pyiron_contrib/blob/5e4106f17e4ad0f5f2899df383e9d8b16753f806/pyiron_contrib/generic/s3io.py#L24).